iothub_client/src/iothubtransport_amqp_common.c

```diff
@@ -2492,57 +2419,49 @@ STRING_HANDLE IoTHubTransport_AMQP_Common_GetHostname(TRANSPORT_LL_HANDLE handle
     return result;
 }
 
-IOTHUB_CLIENT_RESULT IoTHubTransport_AMQP_Common_SendMessageDisposition(MESSAGE_CALLBACK_INFO* message_data, IOTHUBMESSAGE_DISPOSITION_RESULT disposition)
+IOTHUB_CLIENT_RESULT IoTHubTransport_AMQP_Common_SendMessageDisposition(IOTHUB_DEVICE_HANDLE handle, IOTHUB_MESSAGE_HANDLE message_handle, IOTHUBMESSAGE_DISPOSITION_RESULT disposition)
 {
     IOTHUB_CLIENT_RESULT result;
-    if (message_data == NULL)
+    if (message_handle == NULL || handle == NULL)
     {
         /* Codes_SRS_IOTHUBTRANSPORT_AMQP_COMMON_10_001: [If messageData is NULL, IoTHubTransport_AMQP_Common_SendMessageDisposition shall fail and return IOTHUB_CLIENT_INVALID_ARG.] */
-        LogError("Failed sending message disposition (message_data is NULL)");
+        LogError("Failed sending message disposition (handle=%p, message_handle=%p)", handle, message_handle);
         result = IOTHUB_CLIENT_INVALID_ARG;
     }
     else
     {
+        MESSAGE_DISPOSITION_CONTEXT_HANDLE device_message_disposition_info;
+
         /* Codes_SRS_IOTHUBTRANSPORT_AMQP_COMMON_10_002: [If any of the messageData fields are NULL, IoTHubTransport_AMQP_Common_SendMessageDisposition shall fail and return IOTHUB_CLIENT_INVALID_ARG.] */
-        if (message_data->messageHandle == NULL || message_data->transportContext == NULL)
+        // Codes_SRS_IOTHUBTRANSPORT_AMQP_COMMON_09_112: [A DEVICE_MESSAGE_DISPOSITION_INFO instance shall be created with a copy of the `link_name` and `message_id` contained in `message_data`]
+        if (IoTHubMessage_GetDispositionContext(message_handle, &device_message_disposition_info) != IOTHUB_MESSAGE_OK ||
+            device_message_disposition_info == NULL)
         {
-            LogError("Failed sending message disposition (message_data->messageHandle (%p) or message_data->transportContext (%p) are NULL)", message_data->messageHandle, message_data->transportContext);
+            // Codes_SRS_IOTHUBTRANSPORT_AMQP_COMMON_09_113: [If the DEVICE_MESSAGE_DISPOSITION_INFO fails to be created, `IoTHubTransport_AMQP_Common_SendMessageDisposition()` shall fail and return IOTHUB_CLIENT_ERROR]
+            LogError("Invalid message handle (no disposition context found)");
             result = IOTHUB_CLIENT_INVALID_ARG;
```

Comment:
or should this now be `IOTHUB_CLIENT_ERROR`

Reply:
It makes sense for it to be a INVALID_ARG, since this is not expected to happen because of an internal SDK error.
The only situation when this would happen is if the customer creates or clones an IOTHUB_MESSAGE_HANDLE and passes it down to *SendMessageDisposition(...), which would be an invalid arg.